src/data/types.rs

```diff
@@ -281,113 +281,165 @@ pub(super) fn print_type(
+    for (index, declarator_type) in unrolled_type[..unrolled_type.len() - 1].iter().enumerate() {
+        match declarator_type {
+            Array(_, array_type) => {
+                prefixes.push(String::new());
+                postfixes.push(match array_type {
+                    ArrayType::Fixed(length) => format!("[{}]", length),
+                    ArrayType::Unbounded => "[]".to_string(),
+                });
             }
+            Type::Function(function_type) => {
+                prefixes.push(String::new());
 
-            let pointer = if qs != &Default::default() && name != InternedStr::default() {
-                format!("*{} {}", qs, name)
-            } else {
-                format!("*{}{}", qs, name)
-            };
-            write!(f, "{}", pointer)?;
-            if depth {
-                write!(f, ")")?;
+                let params = &function_type.params;
+                let mut buff = String::new();
+                write!(buff, "(")?;
+                for (index, symbol) in params.iter().enumerate() {
+                    let symbol = symbol.get();
+                    write!(buff, "{}", symbol)?;
+                    if index != params.len() - 1 || function_type.varargs {
+                        write!(buff, ", ")?;
+                    }
+                }
+
+                if function_type.varargs {
+                    write!(buff, "...")?;
+                }
+
+                write!(buff, ")")?;
+                postfixes.push(buff);
             }
-            Ok(())
-        }
-        Type::Array(to, _) => print_mid(to, name, f),
-        _ => {
-            if let Some(name) = name {
-                write!(f, " {}", name)?;
+            Pointer(_, qs) => {
+                let needs_parens = match unrolled_type[index + 1] {
+                    Array(_, _) | Function(_) => true,
+                    _ => false,
+                };
+
+                prefixes.push(format!(
+                    "{}*{}",
+                    if needs_parens { "(" } else { "" },
+                    if *qs != Default::default() {
+                        format!("{} ", qs)
+                    } else {
+                        String::new()
+                    }
+                ));
+
+                if needs_parens {
+                    postfixes.push(")".to_string());
+                } else {
+                    postfixes.push(String::new());
+                }
             }
-            Ok(())
+            _ => unreachable!(),
         }
     }
-}
-fn print_post(ctype: &Type, f: &mut Formatter) -> fmt::Result {
-    match ctype {
-        Type::Pointer(to, _) => print_post(to, f),
-        Type::Array(to, size) => {
-            write!(f, "[")?;
-            if let ArrayType::Fixed(size) = size {
-                write!(f, "{}", size)?;
-            }
-            write!(f, "]")?;
-            print_post(to, f)
+
+    let final_type = unrolled_type[unrolled_type.len() - 1];
```

Comment:
I think this would be nicer as `unrolled_type.pop().unwrap()` up at the top and then you iterate normally for the `enumerate` instead of skipping it. Alternatively you could return `(Vec<declarators>, specifier_type)` from `unroll_type`.

Reply:
I was going to change it, but the way I detect whether or not parentheses are needed depends on the vector holding the complete type. I might still change it, but it will require more code than the currently very elegant parenthesis detection.

Reply:
Hmm I see what you mean, it would have to be `unrolled_type.get(i + 1).unwrap_or(specifier_type)`.